File: check.js
// ============================================================
// SESSION SCHEDULER — CHECK SCRIPT
// Node.js version — same logic as check.py, useful for
// GitHub Actions or any environment with Node instead of Python.
// ============================================================

// ============================================================
// CUSTOMIZE THESE VALUES
// ============================================================

// Your Firebase Realtime Database URL (same project as calendar.html)
const FIREBASE_URL = 'https://YOUR_PROJECT_ID-default-rtdb.YOUR_REGION.firebasedatabase.app';

// The public URL where the calendar is hosted (e.g. your Netlify site)
const CALENDAR_LINK = 'https://YOUR_SITE.netlify.app/';

// How long to wait after the last edit before sending a recap
const COOLDOWN_MS = 15 * 60 * 1000;

// How often the "please mark your availability" reminder repeats
const REMINDER_INTERVAL_MS = 3 * 24 * 60 * 60 * 1000;

// Minimum number of people free on a day for it to count as a candidate
const MIN_FREE_FOR_RECAP = 4;

// Minimum number of people free for a day to trigger the big
// "we found a date!" announcement (separate channel, one-time per date)
const ANNOUNCE_THRESHOLD = 6;

// The big announcement message — write your own here
const ANNOUNCEMENT_MESSAGE = "Heads up — we found a shared session date! Time to get the party together.";

// ============================================================

const SCHEDULE_WEBHOOK = process.env.DISCORD_WEBHOOK_SCHEDULE;
const ANNOUNCE_WEBHOOK = process.env.DISCORD_WEBHOOK_ANNOUNCE;

function formatDate(key){
  const [y, m, d] = key.split('-');
  return `${d}.${m}.${y}`;
}

function todayKey(){
  return new Date().toISOString().slice(0, 10);
}

async function getJSON(path){
  const res = await fetch(`${FIREBASE_URL}/${path}.json`, {
    headers: { 'User-Agent': 'SessionScheduler/1.0 (cron script)' }
  });
  return await res.json();
}

async function putJSON(path, value){
  await fetch(`${FIREBASE_URL}/${path}.json`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json', 'User-Agent': 'SessionScheduler/1.0 (cron script)' },
    body: JSON.stringify(value)
  });
}

async function sendWebhook(url, content){
  if(!url){
    console.log('ERROR: webhook URL is missing (env var is empty or misnamed).');
    return;
  }
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'User-Agent': 'SessionScheduler/1.0 (cron script)' },
    body: JSON.stringify({ content })
  });
  if(!res.ok){
    const text = await res.text();
    console.log(`ERROR: Discord rejected the message. Status ${res.status}: ${text}`);
  } else {
    console.log('Webhook sent successfully.');
  }
}

// Note: Firebase silently turns sequential numeric keys (e.g. "1","2","3")
// into a JSON array instead of an object — Object.values() handles both
// shapes fine in JavaScript, so no special-casing is needed here.
function analyzeDays(data){
  const today = todayKey();
  return Object.keys(data)
    .filter(key => key >= today)
    .map(key => {
      const values = Object.values(data[key]);
      return {
        key,
        free: values.filter(v => v === 'free').length,
        busy: values.filter(v => v === 'busy').length
      };
    });
}

function bestDays(days, minFree){
  return days
    .filter(d => d.free >= minFree && d.busy === 0)
    .sort((a, b) => b.free - a.free);
}

function formatList(days){
  if(!days.length) return null;
  return days.map(d => `${formatDate(d.key)} — ${d.free} players`).join('\n');
}

async function runReminder(days){
  const maxFree = days.reduce((max, d) => (d.busy === 0 ? Math.max(max, d.free) : max), 0);
  if(maxFree >= ANNOUNCE_THRESHOLD){
    console.log('No reminder needed, a suitable date has already been found.');
    return;
  }

  const today = todayKey();
  const announced = (await getJSON('announced')) || {};
  const pastSessions = Object.keys(announced).filter(key => key < today).sort();

  if(pastSessions.length){
    const lastSession = pastSessions[pastSessions.length - 1];
    const daysSince = (Date.now() - new Date(lastSession + 'T00:00:00Z').getTime()) / (24 * 60 * 60 * 1000);
    if(daysSince < 3){
      console.log(`Less than 3 days since the last session (${lastSession}), staying quiet for now.`);
      return;
    }
  }

  const lastReminder = await getJSON('meta/last_reminder_time');
  const now = Date.now();
  if(lastReminder && (now - lastReminder) < REMINDER_INTERVAL_MS){
    console.log(`Last reminder was ${Math.round((now - lastReminder) / 3600000)}h ago, not due yet.`);
    return;
  }

  const message = 'Reminder: mark your availability in the calendar so we can find a shared session date!';
  await sendWebhook(SCHEDULE_WEBHOOK, `*${message}*\n\n${CALENDAR_LINK}`);
  await putJSON('meta/last_reminder_time', now);
}

async function runRecap(days, data){
  const lastEditTime = await getJSON('meta/last_edit_time');
  const lastEditor = await getJSON('meta/last_editor');
  const lastNotified = await getJSON('meta/last_notified_edit_time');

  if(!lastEditTime){ console.log('No edits yet.'); return; }
  if(lastNotified === lastEditTime){ console.log('This edit has already been reported.'); return; }

  const elapsed = Date.now() - lastEditTime;
  if(elapsed < COOLDOWN_MS){
    console.log(`Cooldown in progress (${Math.round(elapsed / 1000)}s / ${Math.round(COOLDOWN_MS / 1000)}s).`);
    return;
  }

  if(Object.keys(data).length === 0){
    console.log('The calendar has no entries at all, skipping recap.');
    await putJSON('meta/last_notified_edit_time', lastEditTime);
    return;
  }

  const candidates = bestDays(days, MIN_FREE_FOR_RECAP);
  const total = candidates.length;

  if(total === 0){
    await sendWebhook(SCHEDULE_WEBHOOK,
      `*${lastEditor} made some updates, go check it out!*\n\n${CALENDAR_LINK}`);
    await putJSON('meta/last_notified_edit_time', lastEditTime);
    return;
  }

  const listText = formatList(candidates.slice(0, 10));
  const maxCount = candidates[0].free;

  if(maxCount < ANNOUNCE_THRESHOLD){
    await sendWebhook(SCHEDULE_WEBHOOK,
      `*${lastEditor} made some updates to the calendar!\n\n` +
      `Possible session days: ${total}\n\n${listText}*\n\n${CALENDAR_LINK}`);
  }

  if(maxCount >= ANNOUNCE_THRESHOLD){
    const announced = (await getJSON('announced')) || {};
    const newDate = candidates.find(d => d.free === maxCount && !announced[d.key]);
    if(newDate){
      await sendWebhook(ANNOUNCE_WEBHOOK, `${ANNOUNCEMENT_MESSAGE}\n\n${CALENDAR_LINK}`);
      await putJSON(`announced/${newDate.key}`, true);
    }
  }

  await putJSON('meta/last_notified_edit_time', lastEditTime);
}

async function main(){
  console.log('VERSION: check-js-1');
  const data = (await getJSON('data')) || {};
  const days = analyzeDays(data);

  await runReminder(days);
  await runRecap(days, data);
}

main().catch(err => { console.error(err); process.exit(1); });

File: check.py
#!/usr/bin/env python3
# ============================================================
# SESSION SCHEDULER — CHECK SCRIPT
# Python version — works on any Linux box (e.g. a Raspberry Pi),
# no Node.js or Docker required.
# ============================================================

import json
import os
import sys
import time
import urllib.request
from datetime import datetime, timedelta, timezone

# ============================================================
# CUSTOMIZE THESE VALUES
# ============================================================

# Your Firebase Realtime Database URL (same project as calendar.html)
FIREBASE_URL = 'https://YOUR_PROJECT_ID-default-rtdb.YOUR_REGION.firebasedatabase.app'

# The public URL where the calendar is hosted (e.g. your Netlify site)
CALENDAR_LINK = 'https://YOUR_SITE.netlify.app/'

# How long to wait after the last edit before sending a recap
COOLDOWN_MS = 15 * 60 * 1000

# How often the "please mark your availability" reminder repeats
REMINDER_INTERVAL_MS = 3 * 24 * 60 * 60 * 1000

# Minimum number of people free on a day for it to count as a candidate
MIN_FREE_FOR_RECAP = 4

# Minimum number of people free for a day to trigger the big
# "we found a date!" announcement (separate channel, one-time per date)
ANNOUNCE_THRESHOLD = 6

# The big announcement message — write your own here
ANNOUNCEMENT_MESSAGE = "Heads up — we found a shared session date! Time to get the party together."

# ============================================================

AIKATAULU_WEBHOOK = os.environ.get('DISCORD_WEBHOOK_SCHEDULE')
ANNOUNCE_WEBHOOK = os.environ.get('DISCORD_WEBHOOK_ANNOUNCE')


def format_date(key):
    y, m, d = key.split('-')
    return f'{d}.{m}.{y}'


def today_key():
    return datetime.now(timezone.utc).strftime('%Y-%m-%d')


def now_ms():
    return int(time.time() * 1000)


HEADERS_JSON = {'Content-Type': 'application/json', 'User-Agent': 'SessionScheduler/1.0 (cron script)'}


def get_json(path):
    url = f'{FIREBASE_URL}/{path}.json'
    req = urllib.request.Request(url, headers={'User-Agent': HEADERS_JSON['User-Agent']})
    with urllib.request.urlopen(req, timeout=15) as res:
        raw = res.read().decode('utf-8')
        return json.loads(raw) if raw != 'null' else None


def put_json(path, value):
    url = f'{FIREBASE_URL}/{path}.json'
    data = json.dumps(value).encode('utf-8')
    req = urllib.request.Request(url, data=data, method='PUT', headers=HEADERS_JSON)
    with urllib.request.urlopen(req, timeout=15):
        pass


def send_webhook(url, content):
    if not url:
        print('ERROR: webhook URL is missing (env var is empty or misnamed).')
        return
    data = json.dumps({'content': content}).encode('utf-8')
    req = urllib.request.Request(url, data=data, method='POST', headers=HEADERS_JSON)
    try:
        with urllib.request.urlopen(req, timeout=15) as res:
            if res.status < 200 or res.status >= 300:
                print(f'ERROR: Discord rejected the message. Status {res.status}')
            else:
                print('Webhook sent successfully.')
    except urllib.error.HTTPError as e:
        print(f'ERROR: Discord rejected the message. Status {e.code}: {e.read().decode("utf-8", "ignore")}')


def analyze_days(data):
    # Note: Firebase silently turns sequential numeric keys (e.g. "1","2","3")
    # into a JSON array instead of an object, so both shapes are handled here.
    today = today_key()
    days = []
    for key, entries in (data or {}).items():
        if key < today:
            continue
        if isinstance(entries, dict):
            values = list(entries.values())
        elif isinstance(entries, list):
            values = [v for v in entries if v is not None]
        else:
            values = []
        days.append({
            'key': key,
            'free': values.count('free'),
            'busy': values.count('busy')
        })
    return days


def best_days(days, min_free):
    candidates = [d for d in days if d['free'] >= min_free and d['busy'] == 0]
    candidates.sort(key=lambda d: d['free'], reverse=True)
    return candidates


def format_list(days):
    if not days:
        return None
    return '\n'.join(f"{format_date(d['key'])} — {d['free']} players" for d in days)


def run_reminder(days):
    max_free = max([d['free'] for d in days if d['busy'] == 0], default=0)
    if max_free >= ANNOUNCE_THRESHOLD:
        print('No reminder needed, a suitable date has already been found.')
        return

    today = today_key()
    announced = get_json('announced') or {}
    past_sessions = sorted(k for k in announced.keys() if k < today)

    if past_sessions:
        last_session = past_sessions[-1]
        last_session_dt = datetime.strptime(last_session, '%Y-%m-%d').replace(tzinfo=timezone.utc)
        days_since = (datetime.now(timezone.utc) - last_session_dt) / timedelta(days=1)
        if days_since < 3:
            print(f'Less than 3 days since the last session ({last_session}), staying quiet for now.')
            return

    last_reminder = get_json('meta/last_reminder_time')
    now = now_ms()
    if last_reminder and (now - last_reminder) < REMINDER_INTERVAL_MS:
        hours = round((now - last_reminder) / 3600000)
        print(f'Last reminder was {hours}h ago, not due yet.')
        return

    message = 'Reminder: mark your availability in the calendar so we can find a shared session date!'
    send_webhook(AIKATAULU_WEBHOOK, f'*{message}*\n\n{CALENDAR_LINK}')
    put_json('meta/last_reminder_time', now)


def run_recap(days, data):
    last_edit_time = get_json('meta/last_edit_time')
    last_editor = get_json('meta/last_editor')
    last_notified = get_json('meta/last_notified_edit_time')

    if not last_edit_time:
        print('No edits yet.')
        return
    if last_notified == last_edit_time:
        print('This edit has already been reported.')
        return

    elapsed = now_ms() - last_edit_time
    if elapsed < COOLDOWN_MS:
        print(f'Cooldown in progress ({round(elapsed / 1000)}s / {round(COOLDOWN_MS / 1000)}s).')
        return

    if not data:
        print('The calendar has no entries at all, skipping recap.')
        put_json('meta/last_notified_edit_time', last_edit_time)
        return

    candidates = best_days(days, MIN_FREE_FOR_RECAP)
    total = len(candidates)

    if total == 0:
        send_webhook(AIKATAULU_WEBHOOK,
            f'*{last_editor} made some updates, go check it out!*\n\n{CALENDAR_LINK}')
        put_json('meta/last_notified_edit_time', last_edit_time)
        return

    list_text = format_list(candidates[:10])
    max_count = candidates[0]['free']

    if max_count < ANNOUNCE_THRESHOLD:
        send_webhook(AIKATAULU_WEBHOOK,
            f'*{last_editor} made some updates to the calendar!\n\n'
            f'Possible session days: {total}\n\n{list_text}*\n\n{CALENDAR_LINK}')

    if max_count >= ANNOUNCE_THRESHOLD:
        announced = get_json('announced') or {}
        new_date = next((d for d in candidates if d['free'] == max_count and not announced.get(d['key'])), None)
        if new_date:
            send_webhook(ANNOUNCE_WEBHOOK, f'{ANNOUNCEMENT_MESSAGE}\n\n{CALENDAR_LINK}')
            put_json(f"announced/{new_date['key']}", True)

    put_json('meta/last_notified_edit_time', last_edit_time)


def main():
    print('VERSION: check-py-1')
    data = get_json('data') or {}
    days = analyze_days(data)

    run_reminder(days)
    run_recap(days, data)


if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        print(f'ERROR: {e}', file=sys.stderr)
        sys.exit(1)
